File: Haskell/Básicos/09-2.hs
{- 
	Encontre os 10 últimos anos bissextos (dica: use a função length para determinar o tamanho da lista).
-}
-- Função que verifica se um ano é bissexto
isBissexto :: Int -> Bool
isBissexto ano
    | ano `mod` 400 == 0 = True
    | ano `mod` 100 == 0 = False
    | ano `mod` 4 == 0 = True
    | otherwise = False

-- Função que gera a lista de anos bissextos desde o ano 1 até o ano atual
anosBissextos :: [Int]
anosBissextos = filter isBissexto [1..anoAtual]
  where anoAtual = 2024  -- Substituir por uma função para obter o ano atual, se disponível

-- Função que encontra os 10 últimos anos bissextos
ultimos10AnosBissextos :: [Int]
ultimos10AnosBissextos = reverse (take 10 (reverse anosBissextos))

-- Função principal para exibir os 10 últimos anos bissextos
main :: IO ()
	main = print ultimos10AnosBissextos

File: Programação Funcional/Básicos/Scala/09-2.scala
object Main {
  // Função que verifica se um ano é bissexto
  def isBissexto(ano: Int): Boolean = {
    ano % 400 == 0 || (ano % 100 != 0 && ano % 4 == 0)
  }

  // Função que gera a lista de anos bissextos desde o ano 1 até o ano atual
  def anosBissextos: List[Int] = {
    val anoAtual = 2024 // Substituir por uma função para obter o ano atual, se disponível
    (1 to anoAtual).filter(isBissexto).toList
  }

  // Função que encontra os 10 últimos anos bissextos
  def ultimos10AnosBissextos: List[Int] = {
    anosBissextos.takeRight(10)
  }

  // Função principal para exibir os 10 últimos anos bissextos
  def main(args: Array[String]): Unit = {
    println(ultimos10AnosBissextos)
  }
}
